scripts/casefold_db.py

```diff
@@ -219,35 +219,39 @@ def update_local_associations(
 
         try:
             # Delete each association, and send an email mentioning the affected MXID.
-            if delta.to_delete is not None:
+            if delta.to_delete is not None and not dry_run:
                 for to_delete in delta.to_delete:
-                    if send_email and not dry_run:
+                    if send_email and to_delete.mxid != delta.to_update.mxid:
```

Comment:
I'm not sure I understand the logic here. As per the comment (and the rest of the script), the email we don't want to send an email to is one we don't want to delete, but update instead. So the deletion should also be conditioned on `to_delete.mxid != delta.to_update.mxid`, otherwise how are we meant to update a row that doesn't exist anymore?

Reply:
If I have BOB@x and Bob@x bound to @bob:x, our solution to this is (if I understand):
* Unbind BOB@x
* Rename Bob@x → bob@x
* No e-mail sent: no MXID conflicts

**However we still want to unbind BOB@x**. The `continue` statement in the original code prevents this. I was thinking this causes 'Rename Bob@x → bob@x' to fail (it would at least explain the unique index conflicts) ... but you've made me doubt myself so I'm going to take a moment to trace this out.

Reply:
N.B. In either case, it's still a **massive** bug that the `continue` statements changes the control flow (i.e. whether we do a delete) depending upon we send e-mail or not.

Reply:
I'm convinced that the logic as it stands is incorrect (assuming `send_email`), but the proposed change fixes it.

I believe your misunderstanding is that the row being (proposed to be) deleted and updated is the same; it never is — `delta.to_update` is not present in `delta.to_delete`.

See my reasoning here:
[sydent_casefold_thinking_anno.pdf](https://github.com/matrix-org/sydent/files/7957727/sydent_casefold_thinking_anno.pdf)

* Case (1): If we have to_update=(Bob@x → bob@x) and to_delete=[bob@x], and assume your suggested logic that neither sending an e-mail or a delete should be performed, then the end result is that
Bob@x is changed to bob@x, leading to a unique index violation and a crash (since bob@x was not deleted beforehand).
* Case (2): If we have to_update=(Bob@x → bob@x) and to_delete=[BOB@x], and assume your suggested logic that neither sending an e-mail or a delete should be performed, then the end result is that
Bob@x is changed to bob@x, but BOB@x remains in the database. No crash but this is a logic error because we haven't actually removed all the casefolding duplicates.

Reply:
Oh actually I was misreading/misremembering the code, apologies. For some reason I thought `to_delete` might contain the binding that's also in `to_update`, but of course that'd make no sense :facepalm: